subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesController.java

```diff
@@ -112,135 +115,185 @@ public Map<String, String> getProperties() {
+                fromGradleHome,
+                fromBuildRoot,
+                fromGradleUserHome,
+                fromEnvVariables,
+                fromSystemProperties,
+                fromStartParamProjectProperties
+            );
         }
-    }
-
-    private interface PropertiesState<T extends GradleProperties> {
-        T gradleProperties();
 
-        PropertiesState<T> loadProperties(File dir, boolean setSystemProperties);
+        private void unload() {
+            this.loaded = null;
+        }
     }
 
-    private class NotLoadedBuildProperties implements PropertiesState<MutableGradleProperties> {
 
-        private final BuildIdentifier buildId;
+    private static class LoadedBuildScopedState {
 
-        private NotLoadedBuildProperties(BuildIdentifier buildId) {
-            this.buildId = buildId;
-        }
+        private final File buildRootDir;
+        private final GradleProperties gradleProperties;
+        private final Map<String, String> projectScopedDefaults;
+        private final Map<String, String> projectScopedOverrides;
 
-        @Override
-        public MutableGradleProperties gradleProperties() {
-            throw new IllegalStateException(String.format("GradleProperties for %s have not been loaded yet.", buildId));
+        private LoadedBuildScopedState(
+            File buildRootDir,
+            GradleProperties gradleProperties,
+            Map<String, String> projectScopedDefaults,
+            Map<String, String> projectScopedOverrides
+        ) {
+            this.buildRootDir = buildRootDir;
+            this.gradleProperties = gradleProperties;
+            this.projectScopedDefaults = projectScopedDefaults;
+            this.projectScopedOverrides = projectScopedOverrides;
         }
 
-        @Override
-        public PropertiesState<MutableGradleProperties> loadProperties(File buildRootDir, boolean setSystemProperties) {
-            MutableGradleProperties loadedProperties = propertiesLoader.loadGradleProperties(buildRootDir);
-
-            if (setSystemProperties) {
-                boolean isRootBuild = DefaultBuildIdentifier.ROOT.equals(buildId);
-                systemPropertiesInstaller.setSystemPropertiesFrom(loadedProperties, isRootBuild);
-            }
-
-            Map<String, String> projectProperties = projectPropertiesLoader.loadProjectProperties();
-            loadedProperties.updateOverrideProperties(projectProperties);
-            return new Loaded<>(loadedProperties, buildRootDir);
+        public static LoadedBuildScopedState from(
+            File buildRootDir,
+            Map<String, String> fromGradleHome,
+            Map<String, String> fromBuildRoot,
+            Map<String, String> fromGradleUserHome,
+            Map<String, String> fromEnvVariables,
+            Map<String, String> fromSystemProperties,
+            Map<String, String> fromStartParamProjectProperties
+        ) {
+            ImmutableMap<String, String> defaults = mergeMaps(
```

Comment:
❓ Why not a single hierarchy?

Reply:
If you mean why pre-merge things into `defaults` and `overrides`, then it's because the values from `gradle.properties` in the project directory must be inserted between the defaults and overrides, to preserve the precedance order

Reply:
Ah, I see, `defaults` and `overrides` are preserved. The long constructor line got me fooled.